### PR TITLE
Add mlflow `migrate-filestore` command

### PR DESCRIPTION
**Pull request by harupy**
## 🥞 Stacked PR
Use this [link](https://github.com/mlflow/mlflow/pull/20615/files) to review incremental changes.
- [**stack/fs2db-migration**](https://github.com/mlflow/mlflow/pull/20615) [[Files changed](https://github.com/mlflow/mlflow/pull/20615/files)]
  - [stack/fs2db-docs](https://github.com/mlflow/mlflow/pull/20616) [[Files changed](https://github.com/mlflow/mlflow/pull/20616/files/ac624e4b7f08c7228ac508215c572452982e3a73..b35e34947d08eec38bd98e0c6d5da9d880110375)]

---------
### Related Issues/PRs

#18534
#20463
#20608
#20609

### What changes are proposed in this pull request?

Adds the `mlflow migrate-filestore` CLI command to migrate FileStore (file-based backend) data to a SQLite database. This is the core migration tool for users moving off the legacy FileStore backend.

**New modules under `mlflow/store/fs2db/`:**

- `__init__.py` — Orchestration: creates the target DB, runs all 7 migration phases in a single atomic transaction, prints a summary
- `_helpers.py` — Shared utilities for reading FileStore files (YAML, tags, metrics) using `pathlib.Path`
- `_tracking.py` — Phases 1–6: experiments, runs (params/tags/metrics), datasets, traces, assessments, logged models
- `_registry.py` — Phase 7: registered models, model versions, tags, aliases
- `_verify.py` — Optional `--verify` flag: prints row counts per table from the target DB as a quick sanity check
- `cli.py` — Click command registered as `mlflow migrate-filestore`

**Usage:**

```bash
mlflow migrate-filestore \
  --source /path/to/mlruns \
  --target sqlite:///mlflow.db \
  --verify
```

**Key design decisions:**

- Direct SQLAlchemy ORM inserts (no store API calls) for speed and simplicity
- Single atomic transaction with rollback on failure
- FK-safe insertion order with `session.flush()` between phases
- Version-aware: skips entity types that don't exist on disk (e.g., traces in MLflow 2.x)
- Handles edge cases: NaN/Inf metrics, deleted experiments (`.trash/`), large experiment IDs, V2/V3 trace formats

### How is this PR tested?

- [x] Automated pytest tests (`tests/store/fs2db/test_migration.py`) that generate synthetic FileStore data, run the migration, and compare all entity types (experiments, runs, params, metrics, tags, datasets, traces, assessments, logged models, registered models, model versions, prompts) between source and target via the MLflow public API.
- [x] Manual end-to-end tests against synthetic data generated by both MLflow 2.22.4 and 3.5.1.

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.

Documentation is in the next stacked PR (#20616).

### Does this PR require updating the [MLflow Skills](https://github.com/mlflow/skills) repository?

- [x] No. You can skip the rest of this section.

#### Is this a user-facing change?

- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Added `mlflow migrate-filestore` CLI command to migrate data from the legacy FileStore (file-based) backend to a SQLite database. See #20616 for documentation.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes

#### Should this PR be included in the next patch release?

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/store/fs2db/_tracking.py [hunk 1]
```diff
@@ -0,0 +1,614 @@
...
+    SqlTraceInfo,
+    SqlTraceMetadata,
+    SqlTraceTag,
+)
+
+# ── Phase 1: Experiments + Tags ──────────────────────────────────────────────
+
+
+def migrate_experiments(session: Session, mlruns: Path) -> None:
+    for exp_id in list_experiment_ids(mlruns):
```
Comment (TomeHirata) [inline mlflow/store/fs2db/_tracking.py]: We have many similar logic to retrieve all experiment id from both mlruns and trash dir, can we define a helper function?

File: .github/workflows/fs2db.yml [hunk 2]
```diff
@@ -38,6 +38,23 @@ jobs:
...
+      - name: Verify migration for MLflow ${{ matrix.mlflow-version }}
+        run: |
+          uv run python -c "
+          from sqlalchemy import create_engine, text
+          e = create_engine('sqlite:////tmp/fs2db/${{ matrix.mlflow-version }}/migrated.db')
+          with e.connect() as c:
+              for t in ['experiments','runs','params','metrics','latest_metrics','tags',
+                        'trace_info','logged_models','registered_models','model_versions']:
+                  try:
+                      r = c.execute(text(f'SELECT COUNT(*) FROM {t}')).scalar()
```
Comment (TomeHirata) [inline .github/workflows/fs2db.yml]: Do we plan to add tests?

File: mlflow/store/fs2db/_tracking.py (hunk 1 above)
Comment (harupy) [reply mlflow/store/fs2db/_tracking.py]: Done — extracted `for_each_experiment()` in `_helpers.py` that yields `(exp_dir, exp_id)` from both mlruns and .trash.

File: .github/workflows/fs2db.yml (hunk 2 above)
Comment (harupy) [reply .github/workflows/fs2db.yml]: Added a `--verify` flag to `mlflow migrate-filestore` that counts entities in the source files and compares against DB rows. CI now uses it.

Comment (harupy) [reply .github/workflows/fs2db.yml]: I'll improve the verification process too

File: mlflow/store/fs2db/cli.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    "--target",
+    required=True,
+    help="SQLite URI (e.g. sqlite:///mlflow.db).",
+)
+@click.option(
+    "--verify",
+    is_flag=True,
+    default=False,
+    help="After migration, verify that DB row counts match source file counts.",
+)
```
Comment (harupy) [inline mlflow/store/fs2db/cli.py]: We might not need `--verify` if we have a good test coverage

File: mlflow/store/fs2db/_verify.py [hunk 4]
```diff
@@ -0,0 +1,695 @@
...
+        " FROM inputs i WHERE i.destination_type = 'RUN'"
+        " GROUP BY i.destination_id ORDER BY richness DESC LIMIT 3"
+    ),
+    "run": (
+        "SELECT r.run_uuid,"
+        " (SELECT COUNT(*) FROM params p WHERE p.run_uuid = r.run_uuid)"
+        " + (SELECT COUNT(*) FROM metrics m WHERE m.run_uuid = r.run_uuid)"
+        " + (SELECT COUNT(*) FROM tags t WHERE t.run_uuid = r.run_uuid)"
+        " AS richness"
+        " FROM runs r ORDER BY richness DESC LIMIT 3"
```
Comment (harupy) [inline mlflow/store/fs2db/_verify.py]: This is so ugly (but works)

File: mlflow/store/fs2db/cli.py (hunk 3 above)
Comment (harupy) [reply mlflow/store/fs2db/cli.py]: Simplified the verify logic to only check row counts, and improved the test coverage

File: fs2db/src/generate_synthetic_data.py [hunk 5]
```diff
@@ -280,22 +282,33 @@ def generate_assessments(cfg: SizeConfig, trace_ids: list[str]) -> None:
 
 
 def generate_logged_models(cfg: SizeConfig, experiments: list[ExperimentData]) -> list[str]:
-    """Returns list of model artifact URIs (runs:/<run_id>/model)."""
+    """Returns list of model artifact URIs."""
+    from mlflow.entities.logged_model_input import LoggedModelInput
```
Comment (TomeHirata) [inline fs2db/src/generate_synthetic_data.py:287]: Can we move this import to the module level?

File: fs2db/src/generate_synthetic_data.py [hunk 6]
```diff
@@ -361,8 +374,13 @@ def main() -> None:
     output = os.path.abspath(args.output)
     os.makedirs(output, exist_ok=True)
 
-    os.environ["MLFLOW_TRACKING_URI"] = output
-    mlflow.set_tracking_uri(output)
+    tracking_uri = Path(output).as_uri()
+    os.environ["MLFLOW_TRACKING_URI"] = tracking_uri
```
Comment (TomeHirata) [inline fs2db/src/generate_synthetic_data.py]: q: why do we need both `os.environ["MLFLOW_TRACKING_URI"]=` and `mlflow.set_tracking_uri` though `set_tracking_uri` sets `MLFLOW_TRACKING_URI` internally?

Comment (harupy) [reply fs2db/src/generate_synthetic_data.py]: good catch. we can remove os.environ

Comment (harupy) [reply fs2db/src/generate_synthetic_data.py]:
```suggestion
```

File: mlflow/store/fs2db/__init__.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        _migrate_datasets_for_experiment,
+        _migrate_logged_models_for_experiment,
+        _migrate_one_experiment,
+        _migrate_outputs_for_experiment,
+        _migrate_runs_in_dir,
+        _migrate_traces_for_experiment,
+    )
+    from mlflow.store.fs2db._utils import MigrationStats, for_each_experiment
+
+    def log(msg: str) -> None:
```
Comment (TomeHirata) [inline mlflow/store/fs2db/__init__.py]: Can we move this method outside `migrate`?

Comment (harupy) [reply mlflow/store/fs2db/__init__.py]: Good question. I think that would break mlflow-skinny but let me double check

File: mlflow/store/fs2db/__init__.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+    log("")
+    log("Migration completed successfully!")
+
+    print()
+    print("=" * 50)
+    print("Migration summary:")
+    print("=" * 50)
+    for key, count in stats.items():
+        print(f"  {key}: {count}")
```
Comment (TomeHirata) [inline mlflow/store/fs2db/__init__.py]: nit: I guess this logic can be entirely defined in MigrationStats

File: mlflow/store/fs2db/cli.py (hunk 3 above)
Comment (TomeHirata) [reply mlflow/store/fs2db/cli.py]: +1, I guess this is no longer needed with the current test coverage (but no harm either)

File: tests/store/fs2db/test_migration.py [hunk 9]
```diff
@@ -0,0 +1,265 @@
+import math
+
```
Comment (TomeHirata) [inline tests/store/fs2db/test_migration.py:7]: Could we add an e2e test for the `migrate` method? Basically, it generates data using `generate_synthetic_data` and then verifies that at least all the entity counts on the target database are expected after the migration.

Comment (harupy) [reply tests/store/fs2db/test_migration.py:7]: `migrate` is run in the `clients` fixture. So we could add `--verify` flag there

File: mlflow/store/fs2db/__init__.py [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+                # Release ORM objects from the identity map to keep memory
+                # proportional to one experiment, not the entire FileStore.
+                session.expunge_all()
+            except Exception:
+                session.rollback()
+                log(f"[{i}/{total}] Experiment {exp_id} FAILED — rolled back")
+                raise
+
+        # Model registry is independent of experiments
+        models = list_registered_models(mlruns)
```
Comment (TomeHirata) [inline mlflow/store/fs2db/__init__.py]: Maybe not a realistic case, but do we want to support a case where users use different file paths for tracking URI and registry URI?

File: mlflow/store/fs2db/cli.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+@click.option(
+    "--progress/--no-progress",
+    default=True,
+    help="Show per-experiment progress messages during migration.",
+)
+def migrate_filestore(source: str, target: str, verify: bool, progress: bool) -> None:
+    """Migrate MLflow FileStore data to a SQLite database."""
+    if not target.startswith("sqlite:///"):
```
Comment (TomeHirata) [inline mlflow/store/fs2db/cli.py]: Shall we use existing utils like get_uri_scheme or extract_and_normalize_path?

**Review by TomeHirata — APPROVED** [review]
LGTM once comments are addressed!

File: mlflow/store/fs2db/__init__.py (hunk 10 above)
Comment (harupy) [reply mlflow/store/fs2db/__init__.py]: Good point. I think it is rare. We can support it if requested.

File: tests/store/fs2db/test_migration.py (hunk 9 above)
Comment (TomeHirata) [reply tests/store/fs2db/test_migration.py:7]: Just noticed the fixture is module-scoped, so the existing tests actually have higher coverage than I proposed.

Comment (TomeHirata) [reply tests/store/fs2db/test_migration.py:7]: Iirc, `--verify` doesn't raise, so might not be suitable for tests

File: fs2db/src/generate_synthetic_data.py (hunk 5 above)
Comment (harupy) [reply fs2db/src/generate_synthetic_data.py:287]: This is for mlflow 2.x. LoggedModel doesn't exist in <= 2.x

File: tests/store/fs2db/test_migration.py (hunk 9 above)
Comment (harupy) [reply tests/store/fs2db/test_migration.py:7]: Good call. I've removed `--verify`. I think it's unnecessary. We can always print the row counts.
